### PR TITLE
core: GoogleAIStudio support for tool use

### PR DESCRIPTION
**Pull request by spolu**
## Description

- Add support for getting errors from GoogleAIStudio
- Add support for tool use in GoogleAIStudio

## Risk

Low.

## Deploy Plan

- deploy `core`

### CODE REVIEW

**Review by fontanierh — APPROVED** [review]
Nice 👍  

LGTM assuming this was tested

File: core/src/providers/google_ai_studio.rs [hunk 1]
```diff
@@ -446,32 +525,64 @@ impl LLM for GoogleAiStudioLLM {
...
+                                    None => {
+                                        content = Some(t.clone());
+                                    }
+                                },
+                                None => (),
+                            }
+                            match p.function_call.as_ref() {
+                                Some(fc) => {
+                                    function_calls.push(ChatFunctionCall {
+                                        id: format!("fc_{}", utils::new_id()[0..9].to_string()),
```
Comment (flvndvd) [inline core/src/providers/google_ai_studio.rs:553]: 😢

File: core/src/providers/google_ai_studio.rs [hunk 2]
```diff
@@ -446,32 +525,64 @@ impl LLM for GoogleAiStudioLLM {
...
+        let mut function_calls: Vec<ChatFunctionCall> = vec![];
+
+        match c.candidates {
+            None => (),
+            Some(candidates) => match candidates.len() {
+                0 => (),
+                _ => match &candidates[0].content.parts {
+                    None => (),
+                    Some(parts) => {
+                        for p in parts.iter() {
```
Comment (flvndvd) [inline core/src/providers/google_ai_studio.rs:538]: Not sure to get the value of using a for loop here 🙃

File: core/src/providers/google_ai_studio.rs [hunk 3]
```diff
@@ -610,6 +660,24 @@ pub async fn streamed_chat_completion(
...
+        body["toolConfig"] = json!({
+            "functionCallingConfig": tool_config
+        });
+    }
+
+    if system_instruction.is_some() {
+        body["systemInstruction"] = json!(system_instruction);
+    }
+
+    println!("BODY: {}", body.to_string());
```
Comment (flvndvd) [inline core/src/providers/google_ai_studio.rs:679]: ✂️

File: core/src/providers/google_ai_studio.rs [hunk 4]
```diff
@@ -634,6 +702,7 @@ pub async fn streamed_chat_completion(
                     println!("UNEXPECTED COMMENT");
                 }
                 Some(es::SSE::Event(e)) => {
+                    println!("EVENT DATA: {}", e.data);
```
Comment (flvndvd) [inline core/src/providers/google_ai_studio.rs:705]: ✂️

File: core/src/providers/google_ai_studio.rs [hunk 5]
```diff
@@ -769,89 +860,60 @@ pub async fn streamed_chat_completion(
...
+                        }
+                        match p.function_response.as_ref() {
+                            Some(_) => {
+                                Err(anyhow!("Unexpected function response part in completion"))?;
+                            }
+                            None => (),
                         }
-                        _ => (),
                     }
+                    println!("PARTS: {:?}", parts);
```
Comment (flvndvd) [inline core/src/providers/google_ai_studio.rs:895]: ✂️

**Review by flvndvd — COMMENTED** [review]
\o/

File: core/src/providers/google_ai_studio.rs (hunk 2 above)
Comment (spolu) [reply core/src/providers/google_ai_studio.rs:555]: simpler error reporting. Otherwise I need to return a Result and as a consequence use a map etc...

Comment (fontanierh) [reply core/src/providers/google_ai_studio.rs:555]: Would probably be more readable, more idiomatic, and shorter though

**Comment by spolu** [comment]
Blocked on https://github.com/dust-tt/dust/pull/5497

**Pull request merged**